subprojects/core-api/src/main/java/org/gradle/api/file/CopyProcessingSpec.java

```diff
@@ -54,7 +54,7 @@ public interface CopyProcessingSpec extends ContentFilterable {
      * @param renamer rename function
      * @return this
      */
-    CopyProcessingSpec rename(Transformer<String, String> renamer);
+    CopyProcessingSpec rename(Transformer<@org.jetbrains.annotations.Nullable String, String> renamer);
```

Comment:
Do-able, but I'm not sure what benefit it would bring aside from cleaner imports.

Reply:
It would be good to have only one `@Nullable` annotation to use, with a single option to import... I feel conflicted about this. Why doesn't Java's own `@Nullable` allow type parameters to be annotated? :/

Reply:
I second @lptr's opinion. Could we use meta annotations to create our own `@NullableParameter` annotation instead?

Reply:
We discarded making our own annotation because there are already so many Nullable annotations that each have their own issues and we'd prefer to just use JSpecify's annotations when that's ready. Some tooling already supports JSpecify, but the library isn't production ready yet.

I think our rule for using this is simple. We keep using `@javax.annotation.Nullable` annotations in all places except type parameters. In type parameters, we use the JetBrains one. 

I think we can enforce this with linting. Would that satisfy your concerns?

Reply:
I've pushed some changes here that...

1. Forbid importing org.jetbrains.annotations.*
2. Add ArchUnit rule that prevents using org.jetbrains.annotations on methods

My expectation is that this is temporary until we can switch to JSpecify, which needs Kotlin 1.8.20.

After that, we'd require JSpecify annotations and ban all others.

Reply:
> After that, we'd require JSpecify annotations and ban all others.

Sounds good, @big-guy 👍